### PR TITLE
Ports - Myriam

### PR DESCRIPTION
**Pull request by MyriamWD**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is allowing the communication with the data base.
Describe in your own words what the Controller is doing in Rails | The controller contains the actions that are available to the user. It communicates with the model and view to render a response for the user. 
Describe in your own words what the View is doing in Rails | The view is in charge of formatting the information that is going to be given to the user.
Describe an edge-case controller test you wrote | Edit -> respond with redirect when trying to edit a non-existent task. In this edge-case scenario the user is already in the form page editing a task. While in that process the task got deleted. When the user hits "Save Changes" they will be redirected to tasks_path and inform that the task was deleted using flash error.
What is the purpose of using strong params? (i.e. the params method in the controller) | To control what are the names of the fields and the structure of the parameters. The parameters given need to comply with the criteria in order to be stored in the data base.
How are Rails migrations related to Rails models? | The rails migrations contain instructions in how to change the data base and the model can do that as well. 
Describe one area of Rails that are still unclear on | I'm still struggling to remember what goes where. The level of specificity of some of the elements makes it hard to replicate without looking.

### CODE REVIEW

File: test/controllers/tasks_controller_test.rb [hunk 1]
```diff
@@ -79,24 +76,24 @@
 
       new_task = Task.find_by(name: task_hash[:task][:name])
       expect(new_task.description).must_equal task_hash[:task][:description]
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
-      expect(new_task.completed).must_equal task_hash[:task][:completed]
+      expect(new_task.completion_date).must_equal task_hash[:task][:completion_date]
```
Comment (CheezItMan) [inline test/controllers/tasks_controller_test.rb:79]: Since you don't have a `completion_date` field this should be `expect(new_task.current_status).must_equal task_hash[:task][:current_status]`

And you'll need to update the task_hash.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -1,9 +1,11 @@
 require "test_helper"
 
 describe TasksController do
+  # Note to students:  Your Task model **may** be different and
+  #   you may need to modify this.
   let (:task) {
     Task.create name: "sample task", description: "this is an example for a test",
-                completion_date: Time.now + 5.days
+                completion_date: Time.now # I need to make the completion respond to a toggle
```
Comment (CheezItMan) [inline test/controllers/tasks_controller_test.rb:8]: This should also be `current_status: true` or set it to false.

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
+class TasksController < ApplicationController
+
+  # create tasklist hash
+
+  TASKS = [
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:5]: You can delete this now that you have the model

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    else
+      # For Rails Week 1, not a requirement (not possible right now?) to test this case in controller tests
+      head :not_found
+    end
+  end
+
+  def edit
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      redirect_to tasks_path :flash => {:error => "Task does not exist!"}
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:44]: Nice!

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  def update
+    task = Task.find_by(id: params[:id])
+
+    if task
+      is_successful = task.update(task_params)
+      if is_successful
+        redirect_to task_path(task.id)
+      end
+    else
+      redirect_to tasks_path :flash => {:error => "The task was deleted!"}
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:57]: nice flash message, but also the task might never have existed.

File: app/controllers/tasks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    end
+  end
+
+  def toggle
+    task = Task.find_by(id: params[:id])
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:62]: You should have an `if` statement to see if you can find the task... just saying...

**Comment by CheezItMan** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | Check
Answered comprehension questions | Specifically migrations change the **structure** of the database.  It's normal to struggle to figure out what goes where.  It will take a few weeks.  
Successfully handles: Index, Show | Check
Index & Show tests pass | Check
Successfully handles: New, Create | Check
New & Create tests pass |Almost, see my notes
Successfully handles: Edit, Update |Check
Tests for Edit & Update test valid & invalid task ids |  Almost, see my notes
Successfully handles: Destroy, Task Complete | Check
Tests for Destroy & Task Complete include tests for valid and invalid task ids | Destroy tests work, the mark complete isn't done.
Routes follow RESTful conventions | Check
Uses named routes (like `_path`) | Check
**Overall** | You did well, you missed some of the testing, but minor things.  Overall nice work.  You hit the essentials.
